x/incentives/keeper/grpc_query.go

```diff
@@ -269,6 +270,91 @@ func (q Querier) getGaugeFromIDJsonBytes(ctx sdk.Context, refValue []byte) ([]ty
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "empty request")
+	}
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	pageRes, gauges, err := q.getGaugesWithFilter(ctx, types.KeyPrefixGauges, req.GetPagination(), func(gauge *types.Gauge) bool {
+		return gauge.IsLinkedToPool(req.GetId())
+	})
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	return &types.QueryGaugesByPoolIDResponse{
+		Gauges:     gauges,
+		Pagination: pageRes,
+	}, nil
+}
+
+func (q Querier) InternalGauges(goCtx context.Context, req *types.QueryInternalGaugesRequest) (*types.QueryInternalGaugesResponse, error) {
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "empty request")
+	}
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	pageRes, gauges, err := q.getGaugesWithFilter(ctx, types.KeyPrefixGauges, req.GetPagination(), func(gauge *types.Gauge) bool {
+		return gauge.IsInternalGauge([]string{gammtypes.GAMMTokenPrefix})
+	})
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	return &types.QueryInternalGaugesResponse{
+		Gauges:     gauges,
+		Pagination: pageRes,
+	}, nil
+}
+
+func (q Querier) ExternalGauges(goCtx context.Context, req *types.QueryExternalGaugesRequest) (*types.QueryExternalGaugesResponse, error) {
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "empty request")
+	}
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	pageRes, gauges, err := q.getGaugesWithFilter(ctx, types.KeyPrefixGauges, req.GetPagination(), func(gauge *types.Gauge) bool {
+		return gauge.IsExternalGauge()
+	})
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	return &types.QueryExternalGaugesResponse{
+		Gauges:     gauges,
+		Pagination: pageRes,
+	}, nil
+}
+
+func (q Querier) getGaugesWithFilter(ctx sdk.Context, keyPrefix []byte, pageReq *query.PageRequest, filter GaugesFilterFn) (*query.PageResponse, []types.Gauge, error) {
+	gauges := []types.Gauge{}
+	store := ctx.KVStore(q.Keeper.storeKey)
+	valStore := prefix.NewStore(store, keyPrefix)
+
+	pageRes, err := query.FilteredPaginate(valStore, pageReq, func(key []byte, value []byte, accumulate bool) (bool, error) {
+		// this may return multiple gauges at once if two gauges start at the same time.
+		// for now this is treated as an edge case that is not of importance
+		newGauges, err := q.getGaugeFromIDJsonBytes(ctx, value)
+		if err != nil {
+			return false, err
+		}
+		var hit bool
+		if accumulate {
+			for _, gauge := range newGauges {
+				if !filter(&gauge) {
+					continue
+				}
+				gauges = append(gauges, gauge)
+				hit = true
+			}
+		}
+		return hit, nil
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+	return pageRes, gauges, nil
+}
+
 // filterByPrefixAndDenom filters gauges based on a given key prefix and denom
 func (q Querier) filterByPrefixAndDenom(ctx sdk.Context, prefixType []byte, denom string, pagination *query.PageRequest) (*query.PageResponse, []types.Gauge, error) {
```

Comment:
This function looks not correct if I am right about the way FilteredPaginate works. What do you think @mattverse

Reply:
which part are you sus about?

Reply:
Iterating over newGauges the closure returns false if the denom is not the expected one.
In some situations we can miss a gauge
If len(newGauges) == 2
Index 0 : denom is not expecting one
Index 1 : denom is the expecting one

we are going to exit at first iteration without adding the second gauge in the response.

Reply:
yeah, this is an edge case as mentioned in the comment that we need to fix :(